Math/NumberTheory/Primes/Factorisation/QuadraticSieve.hs

```diff
@@ -0,0 +1,157 @@
+{-# LANGUAGE CPP #-}
+
+module Math.NumberTheory.Primes.Factorisation.QuadraticSieve
+  ( quadraticSieve
+  , findPairs
+  ) where
+
+#if __GLASGOW_HASKELL__ < 803
+import Data.Semigroup
+#endif
+import qualified Data.List as L
+import qualified Data.Vector as V
+import qualified Data.Vector.Mutable as MV
+import qualified Data.IntMap as I
+import qualified Data.IntSet as S
+import qualified Math.NumberTheory.Primes.IntSet as PS
+import Control.Monad
+import Control.Monad.ST
+import Data.Maybe
+import Data.Bifunctor
+import Math.NumberTheory.Roots
+import Math.NumberTheory.Primes
+import Math.NumberTheory.Moduli.Sqrt
+import Math.NumberTheory.Utils.FromIntegral
+
+data SignedPrimeIntSet = SignedPrimeIntSet
+  { sign :: !Bool
+  , primeSet :: !PS.PrimeIntSet
+  } deriving (Show)
+
+data BoolOrPrime = Bool !Bool | PrimeInt !(Prime Int)
+
+insert :: Prime Int -> SignedPrimeIntSet -> SignedPrimeIntSet
+insert prime (SignedPrimeIntSet s ps) = SignedPrimeIntSet s (prime `PS.insert` ps)
+
+nonZero :: SignedPrimeIntSet -> Maybe BoolOrPrime
+nonZero (SignedPrimeIntSet s ps) = case PS.minView ps of
+  Just (prime, _)  -> Just (PrimeInt prime)
+  Nothing          -> if s then Just (Bool True) else Nothing
+
+member :: BoolOrPrime -> SignedPrimeIntSet -> Bool
+member value (SignedPrimeIntSet s ps) = case value of
+  Bool b     -> s == b
+  PrimeInt p -> p `PS.member` ps
+
+xor :: SignedPrimeIntSet -> SignedPrimeIntSet -> SignedPrimeIntSet
+xor (SignedPrimeIntSet s1 ps1) (SignedPrimeIntSet s2 ps2) = SignedPrimeIntSet (s1 /= s2) ((ps1 PS.\\ PS.unPrimeIntSet ps2) <> (ps2 PS.\\ PS.unPrimeIntSet ps1))
+
+-- Given an odd positive composite Integer n and Int parameters b and t,
+-- the Quadratic Sieve attempt to decompose n into smaller factors p and q.
```

Comment:
```suggestion
-- the Quadratic Sieve attempts to decompose n into smaller factors p and q.
```

Reply:
So your return value is either n or p, can you please specify that explicitly.
Also can you specify under which values of n,b,t the algorithm succedes.